### ISSUE TITLE
Check if unit is alive after effect

### ISSUE THREAD

**Issue by olistic** [olistic/warriorjs#63]
A negative effect could potentially kill a unit. We need to check if the unit is still alive after [executing each effect](https://github.com/olistic/warriorjs/blob/6b90d0629e563ea664e822acf3914353c99c5a7c/packages/warriorjs-core/src/Unit.js#L98) and return early if the unit died. Example: a "poisoned" effect (it doesn't exist, at least yet) that damages the unit turn by turn.

**Comment by PB020** [comment]
Hello, I am relatively new to working on with JavaScript on this level. That being said, this should fix this problem, unless I am mistaken.

![heal correction - capture](https://user-images.githubusercontent.com/32374905/40285434-e76d4e98-5c69-11e8-97eb-a608f5c71b39.PNG)

**Comment by PB020** [comment]
No, wait. That would not completely solve the problem. Sorry for the worthless comment.

**Comment by glneto** [comment]
I took a look at this issue, but I imagine that if an effect that damages the unit is created, it would call the unit.takeDamage which already verifies if the unit died. Am I missing something here? Or maybe I misunderstood the issue.

**Comment by olistic** [comment]
I think this will be clearer if I explain [how the turns are played](https://github.com/olistic/warriorjs/blob/56e1cb08247b832fa3538d1cd6c0effcd9ab4d49/packages/warriorjs-core/src/Level.js#L47-L48) by the units in the floor:

1. In the order the units were added to the floor, their `prepareTurn` method is executed. This is when the `playTurn` method of each unit is called, executing the senses immediately but deferring the action for later. **Nothing changes in the game at this point; units just analyze their environment to decide what the best course of action is.**
2. In that same order, the `performTurn` method of each unit is executed. This is when the actions that were deferred for later are executed, having impact in the game.

Example:

```
╔════════╗
║   @s  >║
╚════════╝
```

> We have the Warrior there, next to a Sludge. The warrior is full health, but the sludge is about to die (one hit from the Warrior and it's dead). During (1), both the Warrior and the Sludge decide to attack. Later in (2), the Warrior attacks first, killing the Sludge. The Sludge had already decided that it was going to attack, but it wouldn't be ok if he was able to do so because he's now dead. That's why [this check](https://github.com/olistic/warriorjs/blob/56e1cb08247b832fa3538d1cd6c0effcd9ab4d49/packages/warriorjs-core/src/Unit.js#L134) exists.

---

Now, what this issue is about is putting a similar check in place for the effects. The effects are executed [here](https://github.com/olistic/warriorjs/blob/56e1cb08247b832fa3538d1cd6c0effcd9ab4d49/packages/warriorjs-core/src/Unit.js#L135), during `performTurn`, and can have an impact on the game (e.g. kill the affected unit).

Example:

> Our Warrior is about to die (1 HP) and he's under the Poison effect, that subtracts 2 HP each turn. The poison takes effect [here](https://github.com/olistic/warriorjs/blob/56e1cb08247b832fa3538d1cd6c0effcd9ab4d49/packages/warriorjs-core/src/Unit.js#L135), killing the Warrior. It wouldn't be ok if the Warrior was able to the action after, because he's now dead. So another `isAlive()` check is needed.

**Open questions:**

* Where do we put the `isAlive()` check?
  a. Once after executing the `passTurn` of each of the effects
  b. Once after executing the `passTurn` of all the effects

* Do we want to execute the effects before the action (as it's currently being done) or after?

**Comment by PB020** [comment]
Thanks for that response, it really cleared things up. The `isAlive()` check should occur after the execution of each effect, checking if the current health of the player is at or greater than 0. Unless I am wrong, the performTurn method should be written like this for proper implementation.

```
performTurn() {
    if (this.isAlive()) {
        this.effects.forEach(effect => effect.passTurn());
	if(this.isAlive()){
            if (this.turn.action && !this.isBound()) {
                const [name, args] = this.turn.action;
                this.abilities.get(name).perform(...args);
            }
	}
	else {
	    this.log(`You have $(this.health) out of $(this.maxHealth). You are dead.`);
	}
    }
}
```

**Comment by olistic** [comment]
Hey @PB020, I encourage you to open a PR if you want to try a fix for this! We can discuss the code there with better instruments.

Here are some comments regarding what you already posted though:

```js
this.effects.forEach(effect => effect.passTurn());
if (this.isAlive()) {
```

There you're only checking once, after all effects have been executed. In order to check after each effect, the check needs to be done inside the `forEach`. But to be able to return early, the `forEach` needs to be replaced by a `for..of` loop.

```js
else {
  this.log(`You have $(this.health) out of $(this.maxHealth). You are dead.`);
}
```

This code is redundant and should be removed. If an effect kills the unit, this will be logged by the effect already.

**Comment by IceDune** [comment]
@olistic , @PB020 , can i work on this?

**Comment by olistic** [comment]
@IceDune Sure! Let me know if you have any questions.

**Comment by glneto** [comment]
It's been a while since the last comment on this, so I opened #250 - hope it works as expected. If there is any case I missed let me know.
